Allow to set axes config `visDomain` to any Domain

t20100 opened the pull request:
This PR allows to set the `visDomain` freely.
Beforehand, the `visDomain`s needed to be `[0, number of image pixel]`

This is done by changing the hooks used to retrieve the visible domains and resolution.

closes #987

t20100 commented on apps/storybook/src/TiledHeatmap.stories.tsx:
```diff
@@ -116,23 +116,28 @@ class MandelbrotTilesApi extends TilesApi {
   }
 }
+
+const X_DOMAIN: Domain = [-2, 1];
+const Y_DOMAIN: Domain = [-1.5, 1.5];
```
Alternative was to use `api.xDomain` and `api.yDomain` from the `MandelbrotTilesApi`, but it requires to override the props interface provided for the `Template`.

loichuder replied:
I don't understand. You can use `api.xDomain` instead of `X_DOMAIN` inside the `Template` without changing the `props`, right?

Also, perhaps it is worth having a different story for non-pixel domain.

t20100 replied:
> Template without changing the props, right?
Apparently no because of the typing `api.xDomain` is not part of `TilesApi`

loichuder replied:
> Apparently no because of the typing `api.xDomain` is not part of `TilesApi`

Oh yes true. Didn't realize that `xDomain` and `yDomain` were implemented by `MandelbrotTilesApi`

t20100 commented on packages/lib/src/vis/scaleGamma.ts:
```diff
@@ -10,11 +10,7 @@ import type {
-  const extent = b - a;
-  return (x: number) => (extent !== 0 ? (x - a) / extent : 0.5);
-}
+import { normalize } from './utils';
```
Moved so that I can also use it.

t20100 commented on packages/lib/src/vis/tiles/TiledHeatmap.tsx:
```diff
@@ -23,14 +23,20 @@ function TiledHeatmap(props: Props) {
+  const { baseLayerIndex, baseLayerSize } = api;
+  const { width, height } = baseLayerSize;
+
+  const itemPerPixel = Math.max(
```
This is now the resolution given as the number of image array items per screen pixel

loichuder commented on packages/lib/src/vis/tiles/hooks.ts:
```diff
@@ -1,13 +1,34 @@
+
+  return {
+    xNormVisibleDomain: xVisibleDomain.map((v) =>
+      clamp(normalize(...abscissaVisDomain)(v), 0, 1)
```
Suggestion: you could use a `ScaleLinear` here instead of normalizing "by hand"

t20100 replied:
Yes, why not, I used visx's scaleLinear

t20100 commented on packages/lib/src/vis/tiles/hooks.ts:
```diff
@@ -1,13 +1,32 @@
+import { useAxisSystemContext } from '../shared/AxisSystemContext';
 
-export function useDataPerPixel() {
+export function useScaledVisibleDomains(
```
Following our discussion, I still kept a hook since it is needed both in `TiledHeatmap` and `TiledLayer`

t20100 commented on packages/lib/src/vis/tiles/hooks.ts:
```diff
@@ -1,13 +1,32 @@
   return {
-    xDataPerPixel: (xVisibleDomain[1] - xVisibleDomain[0]) / width,
-    yDataPerPixel: (yVisibleDomain[1] - yVisibleDomain[0]) / height,
+    xVisibleDomain: xVisibleDomain.map(xScale) as Domain,
+    yVisibleDomain: yVisibleDomain.map(yScale) as Domain,
```
Is using same naming for output as `useVisibleDomains` ok for you or best use a different naming (e.g., `xScaledVisibleDomain`... but that starts to be long)?

axelboc replied:
I think it's fine.

loichuder replied:
Agreed.

t20100 commented:
I updated and use a single `useScaledVisibleDomains` hook (use from 2 components).
I also changed the parameters of `getTileOffsets` from rectangle corners to ranges to spare the conversion from domain to corners in `TiledLayer`: commit dbd1c11c12b6baa54ebb160ed19c852b76a1ad27. If you don't think it simplifies the reading, I'll remove it.

loichuder commented on apps/storybook/src/TiledHeatmap.stories.tsx:
```diff
@@ -116,23 +116,29 @@ class MandelbrotTilesApi extends TilesApi {
+      visRatio={Math.abs(
+        (abscissaDomain[1] - abscissaDomain[0]) /
+          (ordinateDomain[1] - ordinateDomain[0])
+      )}
```
The ratio you want to keep should not be `api.baseLayerSize.width / api.baseLayerSize.height` ?

In any case, I think the story would allow to catch more regressions if the span of the abscissa domain was different from the span of the ordinate domain (so far, both span 3).

t20100 replied:
> The ratio you want to keep should not be api.baseLayerSize.width / api.baseLayerSize.height ?

I expect most of the time, but maybe not... But it behaves poorly if there is a large difference between the axes ratio and the ratio of the image since the level of details to then differs between the axes.

> I think the story would allow to catch more regressions if the span of the abscissa domain was different from the span of the ordinate domain

:+1:

loichuder commented on packages/lib/src/vis/tiles/TiledLayer.tsx:
```diff
@@ -16,28 +17,21 @@ interface Props extends ColorMapProps {
+  const { xVisibleDomain, yVisibleDomain } = useScaledVisibleDomains(
+    [0, width],
+    [0, height]
+  );
```
Same here, you could pass `api.layerSizes[layer]` directly

loichuder reviewed (approved):
This is way clearer in my opinion. Great job :+1:
